odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+        ...     return vol
+        >>> ndim = 3
+        >>> vsize = 100
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> A = DiscreteGradient(disc, (1.,)*ndim, zero_padding=True)
+        >>> Af = A(f)
+        >>> g = A.range.one()
+        >>> Adg = A.adjoint(g)
+        >>> g.inner(Af) - f.inner(Adg)
+        0.0
+        >>> B = DiscreteDivergence(disc, (1.,)*ndim, zero_padding=True)
+        >>> Bg = B(g)
+        >>> Bdf = B.adjoint(f)
+        >>> f.inner(Bg) - g.inner(Bdf)
+        0.0
+        >>> ndim = 3
+        >>> vsize = 9
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> A = DiscreteGradient(disc, (1.,)*ndim, zero_padding=True)
+        >>> Af = A(f)
+        >>> g = A.range.one()
+        >>> Adg = A.adjoint(g)
+        >>> g.inner(Af) - f.inner(Adg)
+        0.0
+        >>> B = DiscreteDivergence(disc, (1.,)*ndim, zero_padding=True)
+        >>> Bg = B(g)
+        >>> Bdf = B.adjoint(f)
+        >>> f.inner(Bg) - g.inner(Bdf)
+        0.0
+
+        """
+        rhs_data = np.asanyarray(rhs)
+        ndim = rhs_data.ndim
+
+        dx = self.voxel_size
+        if np.size(dx) == 1:
+            dx = [dx for _ in range(ndim)]
+
+        for axis in range(ndim):
+            out[axis][:] = discrete_part_deriv(rhs_data, axis, dx[axis],
+                                               self.edge_order,
+                                               self.zero_padding)
+
+    @property
+    def adjoint(self):
+        """Assuming zero padding, this returns the adjoint operator given by
+        the negative of the `DiscreteDivergence` operator.
+
+        Note that the first argument (`space`) of the `DiscreteDivergence`
+        operator is not the range but the  domain of the `DiscreteGradient`
+        operator.
+        """
+        return -DiscreteDivergence(self.domain, voxel_size=self.voxel_size,
+                                   edge_order=self.edge_order,
+                                   zero_padding=self.zero_padding)
+
+
+# noinspection PyAbstractClass
+class DiscreteDivergence(Operator):
+    """Divergence operator for any number of dimensions.
+
+    Calls helper function `discrete_part_deriv` for each component of the
+    input product space vector vector. For the adjoint of the
+    `DiscreteDivergence` operator to match the negative `DiscreteGradient`
+    operator requires `zero_padding`.
+    """
+
+    def __init__(self, space, voxel_size=(1,), edge_order=2,
+                 zero_padding=True):
+        """Initialize a `DiscreteDivergence` operator instance.
+
+        Zero padding is required for the negative `DiscreteGradient`
+        operator to match the adjoint of the `DiscreteDivergence` operator.
```

Comment:
Yes, without zero padding the surface term after partial integration won't be zero any more and has to be handled properly.